### PR TITLE
new(engine): add selective rule overrides

### LINKED ISSUE #1340: Support partial overrides / merging  for existing rule (e.g. `priority` field) to complement the append functionality
**Motivation**

Falco has around 60 default rules configured. However, I am only interested in say configuring 15 rules with an alert policy because of my setup. `Falcosidekick` is a great tool for managing the alerting of falco rules with setting a default priority on what it should fire. 

At current, there are priorities of notice, informational, error (which I find a weird name), but also the priorities `alert, critical and emergency`. As part of the default rules, `alert` and `emergency` are not used. Weird. 

My main goal is to change 15 alerts to the `alert` priority.

The only way to do this is **copy the entire rule** into my `rules.local.yaml`. There has to be a smarter and better way for changing this.

**Feature**
```
- rule: xyz
  append: priority
  priority: alert
```

**Alternatives**

My alternative is copying the entire rule from the default rules into my `rules.local`

**Additional context**

Others have had this problem: https://kubernetes.slack.com/archives/CMWH3EH32/p1596015898189000

### PR DESCRIPTION
**Pull request by LucaGuerra**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area engine

/area tests

**What this PR does / why we need it**:

This is a rather straightforward implementation of https://github.com/falcosecurity/falco/issues/1340#issuecomment-1710135769 . See the comment for examples.

It is now possible to use the `override` key in rules, lists and macros. Some fields can be overridden: for lists and macros there is only one field that we care about, while for rules, the following fields have been selected:
* Available for `append`: `{"condition", "output", "desc", "tags", "exceptions"}`
* Available for `replace`: `{"condition", "output", "desc", "priority", "tags", "exceptions", "enabled", "warn_evttypes", "skip-if-unknown-filter"}`

As per the discussion linked above, it is an error to specify both `append: true` and any `override`. Given the syntax it is not possible to attempt to both replace and append the same field in the same entry.

Also, if you specify the `override` key you *must* specify all the fields that you wish to override and only those fields.

**Which issue(s) this PR fixes**:

Fixes #1340

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new(engine): add selective overrides for Falco rules
```

Also, this needs to be documented ( cc @mikegcoleman )

### CODE REVIEW

**Comment by incertum** [comment]
Thank you @LucaGuerra could you let me know when it would be ready to test drive it?

File: userspace/engine/rule_loader_collector.cpp
Comment (LucaGuerra) [inline userspace/engine/rule_loader_collector.cpp:194]: Note for the reviewer: this string change is making our testing pipelines fail. No big deal, once we pick the right string I can update the pipelines.

File: userspace/engine/rule_loader_collector.cpp
Comment (LucaGuerra) [inline userspace/engine/rule_loader_collector.cpp:58]: Note for the reviewer: the addition of `replace_info` could be cargo cult programming, I am not sure if we need to do more here or if this is useful.

File: userspace/engine/rule_loader.h
Comment (LucaGuerra) [inline userspace/engine/rule_loader.h:461]: Note for the reviewer: while we could add `std::optional`s to the existing structures, those are also used to evaluate and handle actual rules which must have those fields defined. I wouldn't want the rest of the engine to care about whether a mandatory field was defined or not (as it must be) so it sounds ok to keep those fields here.

**Comment by LucaGuerra** [comment]
@incertum if you are able to test this it would be appreciated :)

File: userspace/engine/rule_loader_collector.cpp
Comment (LucaGuerra) [reply userspace/engine/rule_loader_collector.cpp:58]: After speaking with @jasondellaluce , this seems the right approach.

File: unit_tests/engine/test_rule_loader.cpp
Comment (incertum) [inline unit_tests/engine/test_rule_loader.cpp:7]: Very neat setup @LucaGuerra much more elegant :heart: than what I had tried in the past :woman_facepalming:

File: unit_tests/engine/test_rule_loader.cpp
Comment (incertum) [inline unit_tests/engine/test_rule_loader.cpp:201]: Few ideas to add more tests to gain even more confidence:

Inline with original requirements:

> It is now possible to use the override key in rules, lists and macros. Some fields can be overridden: for lists and macros there is only one field that we care about, while for rules, the following fields have been selected:

Add a test to verify an error when using `append` for the field `priority`

> - Available for append: {"condition", "output", "desc", "tags", "exceptions"}

Add a test for an unsupported field, e.g. `enabled` in `append` mode and verify error message and no success.

> -Available for replace: {"condition", "output", "desc", "priority", "tags", "exceptions", "enabled", "warn_evttypes", "skip-if-unknown-filter"}

> As per the discussion linked above, it is an error to specify both append: true and any override. Given the syntax it is not possible to attempt to both replace and append the same field in the same entry.

Test the error when using both `append` and `override`

> Also, if you specify the override key you must specify all the fields that you wish to override and only those fields.

Create few tests with a mismatch, that is, you use a field in `override` that you have not provided a statement and vice versa add a statement for which there is no matching `override` definition. Actually what is the behavior in that case? Just keep the old and no error?


More ideas:

- Test successive overrides of the same rule just to be future proof and verify the final result.
- Plus I keep calling it out: What is the behavior when I disable a rule using this selective override but then enable all rules with a certain tag and unfortunately the rule I disabled with selective overrides happened to have that tag. Is that rule at the end enabled or disabled? 
- Test error when defining a rule with `override` feature, but the rule was not defined in the first place.

**Review by incertum — COMMENTED** [review]
First pass review focused on the tests. Could we add some more? See comment.

Code implementation LGTM and @jasondellaluce already approved the approach as well :smile:.

File: unit_tests/engine/test_rule_loader.cpp
Comment (LucaGuerra) [reply unit_tests/engine/test_rule_loader.cpp:7]: I noticed we actually do something similar in at least three different locations in the test suite (including this one). This is out of scope for this PR but we need to consolidate them and have one fixture to allow loading engine + rulesets.

File: userspace/engine/rule_loader_collector.cpp
Comment (LucaGuerra) [reply userspace/engine/rule_loader_collector.cpp]: I have updated the code to revert the error messages, I really want to be sure that our tests are happy before merging. Messages can be updated in a subsequent PR coordinated with testing.

File: unit_tests/engine/test_rule_loader.cpp
Comment (LucaGuerra) [reply unit_tests/engine/test_rule_loader.cpp:203]: Good point. I have added error testcases for most cases you mentioned, those are not only useful for these specifics but are easy enough to understand so they can be extended for more cases we can think of

> Actually what is the behavior in that case? Just keep the old and no error?
The way that makes more sense to me is that is an error to ask for an override and not define it and vice versa. Actually, the opposite was not checked and I now added a check for it. Essentially, if you define a field and not the corresponding append/replace override it will result in an error; this is because if you defined the field in an entry with an override section you probably wanted to override that. I also added the corresponding test.

Also, for more complex cases (e.g. enabling/disabling with tags) I think they may be out of scope for this PR and if you can think about unit or integration tests for those cases it would be great to add them.

**Comment by incertum** [comment]
> Also, for more complex cases (e.g. enabling/disabling with tags) I think they may be out of scope for this PR and if you can think about unit or integration tests for those cases it would be great to add them.

Agreed, just wanted to mention it once again :upside_down_face:

File: unit_tests/engine/test_rule_loader.cpp
Comment (incertum) [reply unit_tests/engine/test_rule_loader.cpp:7]: Big +1, it would be super useful to do this cleanup!

File: unit_tests/engine/test_rule_loader.cpp
Comment (incertum) [inline unit_tests/engine/test_rule_loader.cpp:261]: [nit] `Key 'priority' cannot be appended to, use 'replace' instead` -> same for `enabled`

File: unit_tests/engine/test_rule_loader.cpp
Comment (incertum) [inline unit_tests/engine/test_rule_loader.cpp:304]: [nit] `you are overriding a rule, but no rule by that name already exists` or similar

File: unit_tests/engine/test_rule_loader.cpp
Comment (incertum) [inline unit_tests/engine/test_rule_loader.cpp:350]: [nit] same can we make the error message more expressive?

File: unit_tests/engine/test_rule_loader.cpp
Comment (LucaGuerra) [reply unit_tests/engine/test_rule_loader.cpp:304]: This is an "error message contains" not equals :)

Comment (incertum) [reply unit_tests/engine/test_rule_loader.cpp:304]: :)

File: unit_tests/engine/test_rule_loader.cpp
Comment (LucaGuerra) [reply unit_tests/engine/test_rule_loader.cpp]: Good point, updated!

**Review by incertum — APPROVED** [review]
/approve

Wonderful work @LucaGuerra, thank you!

**Review by jasondellaluce — APPROVED** [review]
/approve

/hold

Super job Luca! I'm approving and holding just so that you can decide when you feel confident in merging this, but it looks great to me.

**Comment by incertum** [comment]
/unhold

Had spoken with Luca on slack. We are good to merge it.

**Comment by incertum** [comment]
The following 2 issues were duplicates of the primary issue https://github.com/falcosecurity/falco/issues/1340:

- https://github.com/falcosecurity/falco/issues/2477
- https://github.com/falcosecurity/falco/issues/2541

Link them for tracking.

**Comment by mikegcoleman** [comment]
Added a docs issue to track the documentation need

https://github.com/falcosecurity/falco-website/issues/1226

**Comment by pmusa** [comment]
I think this part also needs to be updated:
https://falco.org/docs/rules/controlling-rules/#via-custom-rule-definition

**Pull request merged**
